runtime-sdk/src/storage/confidential.rs

```diff
@@ -0,0 +1,617 @@
+use std::convert::TryInto as _;
+
+use anyhow;
+use hmac::{Hmac, Mac as _, NewMac as _};
+use sha2::Sha512Trunc256;
+use slog::error;
+use thiserror::Error;
+use zeroize::{Zeroize, Zeroizing};
+
+pub use crate::core::common::crypto::mrae::deoxysii::KEY_SIZE;
+use crate::{
+    core::{
+        common::crypto::{
+            hash::Hash,
+            mrae::deoxysii::{self, NONCE_SIZE},
+        },
+        storage::mkvs,
+    },
+    storage::Store,
+};
+
+type Nonce = [u8; NONCE_SIZE];
+type Kdf = Hmac<Sha512Trunc256>;
+
+/// Unpack the concatenation of (nonce || byte_slice) into (Nonce, &[u8]).
+fn unpack_nonce_slice<'a>(packed: &'a [u8]) -> Option<(&'a Nonce, &'a [u8])> {
+    if packed.len() <= NONCE_SIZE {
+        return None;
+    }
+    let nonce_ref: &'a Nonce = packed[..NONCE_SIZE]
+        .try_into()
+        .expect("nonce size mismatch");
+    Some((nonce_ref, &packed[NONCE_SIZE..]))
+}
+
+/// Errors emitted by the confidential store.
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("corrupt key")]
+    CorruptKey,
+
+    #[error("corrupt value")]
+    CorruptValue,
+
+    #[error("decryption failure: {0}")]
+    DecryptionFailure(anyhow::Error),
+}
+
+/// A key-value store that encrypts all content with DeoxysII.
+pub struct ConfidentialStore<S: Store> {
+    inner: S,
+    deoxys: deoxysii::DeoxysII,
+    base_value_prefix: Vec<u8>,
+    nonce_counter: usize,
+    nonce_key: Zeroizing<Vec<u8>>,
+}
+
+impl<S: Store> ConfidentialStore<S> {
+    /// Create a new confidential store with the given keypair.
+    pub fn new_with_key(inner: S, key: [u8; KEY_SIZE], value_context: &[&[u8]]) -> Self {
+        let actual_key = Zeroizing::new(key);
+
+        // Derive a nonce key for nonces used to encrypt storage keys in the store:
+        // nonce_key = KDF(key)
```

Comment:
Hm, HMAC as a KDF isn't amazing, but it's probably ok?  This is only used for nonce derivation right?